lib/web/app/handler.go

```diff
@@ -541,61 +542,107 @@ const (
+//     - /web/launch/some-domain.com?path=%2Farbitrary%2Fpath&query=foo%3Dbar%26baz%3Dqux
 //
-// The original path and query cannot be formed as `web/launch/<original URL>`
-// because `web/launch` route can differ depending on how the user hits the app
-// endpoint:
-//  1. /web/launch/:fqdn/:clusterID/:publicAddr?/:arn?
-//     This route is formed when user clicks on the web UI's app launcher
-//     button from the application listing screen. The app can be directly
-//     resolved since we are able to determine the app's cluster name,
-//     public address, and AWS role name (if defined).
-//  2. /web/launch/<fqdn>?path=<encoded path>&query=<encoded query>
-//     This route is formed when a user hits the app endpoint outside of
-//     the web UI (clicking from a link or copy/pasta link), and the app will
-//     have to be resolved by the fqdn.
+//  2. Building off from previous step, the web app launcher can now redirect the user
+//     to the apps "x-teleport-auth" endpoint to start the auth exchange:
+//     https://some-domain.com/x-teleport-auth?path=%2Farbitrary%2Fpath&query=foo%3Dbar%26baz%3Dqux
 //
-// Isolating the `fqdn` prevents confusing the rest of the param reserved for
-// clusterId, publicAddr, and arn (where the non-query param values are used to
-// create app session). The `web/launcher` will reconstruct the original
-// app URL when ready to redirect the user to the requested endpoint.
-func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string) string {
-	// Note that r.URL.Path field is stored in decoded form where:
-	//  - `/%47%6f%2f` becomes `/Go/`
-	//  - `siema%20elo` becomes `siema elo`
-	// And QueryEscape() will encode spaces as `+`
-	//
-	// QueryEscape is used on the `r.URL.Path` since it is being placed
-	// into the query part of the URL.
-	query := fmt.Sprintf("path=%s", url.QueryEscape(r.URL.Path))
-	if len(r.URL.RawQuery) > 0 {
-		query = fmt.Sprintf("%s&query=%s", query, url.QueryEscape(r.URL.RawQuery))
+//     We will need to reconstruct the same URL ^ along with the stateToken created
+//     by server:
+//     - /web/launch/some-domain.com?path=%2Farbitrary%2Fpath&query=foo%3Dbar%26baz%3Dqux&state=ABCD
+//
+// The URL's are formed this way to help isolate the path params reserved for the app
+// launchers route, where order and existence of previous params matter for this route.
+func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string, req launcherURLParams) string {
+	u := url.URL{
+		Scheme: "https",
+		Host:   proxyPublicAddr,
+		Path:   fmt.Sprintf("/web/launch/%s", hostname),
 	}
 
-	u := url.URL{
-		Scheme:   "https",
-		Host:     proxyPublicAddr,
-		Path:     fmt.Sprintf("/web/launch/%s", hostname),
-		RawQuery: query,
+	// Presence of a stateToken means we are beginning an app auth exchange.
+	if req.stateToken != "" {
+		u.RawQuery = fmt.Sprintf("state=%s&path=%s", url.QueryEscape(req.stateToken), url.QueryEscape(req.path))
```

Comment:
Similar comment - what you have is fine (you were careful to properly escape the values), but why not use `url.Values` to encode the string rather than sprintf?

```
var v url.Values
v.Add("state", req.stateToken)
v.Add("path", req.path)
u.RawQuery = v.Encode()
```